### ISSUE TITLE
베이스모델_Vision Transformer(ViT)

### ISSUE THREAD

**Issue by shualoalumin** [seungboAn/National_OceanoGraphic#16]

**Comment by leehan32** [comment]
pip install transformers 설치해야함

import json
from keras.applications import ResNet50, InceptionV3, MobileNetV2, VGG16
from keras.models import Model
from keras.layers import Dense, GlobalAveragePooling2D
from keras.preprocessing.image import ImageDataGenerator
from keras.optimizers import Adam
from transformers import ViTForImageClassification, ViTFeatureExtractor
import tensorflow as tf

# 데이터 경로 설정
train_dir = r"C:\Users\dkdle\Desktop\archive\Train_Test_Valid\Train"
val_dir = r"C:\Users\dkdle\Desktop\archive\Train_Test_Valid\valid"
batch_size = 32
input_shape = (224, 224, 3)

# 데이터 로더 (증강 제거, 단순 리스케일)
data_gen = ImageDataGenerator(rescale=1./255)

train_generator = data_gen.flow_from_directory(
    train_dir,
    target_size=(224, 224),
    batch_size=batch_size,
    class_mode='categorical'
)

val_generator = data_gen.flow_from_directory(
    val_dir,
    target_size=(224, 224),
    batch_size=batch_size,
    class_mode='categorical'
)

# 모델 생성 함수
def create_model(base_model, num_classes):
    x = base_model.output
    x = GlobalAveragePooling2D()(x)
    outputs = Dense(num_classes, activation='softmax')(x)
    return Model(inputs=base_model.input, outputs=outputs)

# 이미지 트랜스포머 모델 추가
def create_vit_model(num_classes, input_shape):
    # Load ViT model and feature extractor
    feature_extractor = ViTFeatureExtractor.from_pretrained("google/vit-base-patch16-224-in21k")
    vit_model = ViTForImageClassification.from_pretrained("google/vit-base-patch16-224-in21k")

    # Wrap it into a TensorFlow model
    inputs = tf.keras.Input(shape=input_shape)
    x = tf.image.resize(inputs, (224, 224))  # Ensure correct input size
    x = tf.keras.layers.Rescaling(1./255)(x)  # Rescale
    outputs = vit_model(inputs=x)["logits"]  # Use logits from ViT
    vit_model = tf.keras.Model(inputs, outputs)

    # Add final dense layer for the specific task
    final_output = Dense(num_classes, activation="softmax")(outputs)
    model = Model(inputs=inputs, outputs=final_output)

    return model

# 히스토리 저장 함수
def save_history(history, filename):
    with open(filename, 'w') as f:
        json.dump(history.history, f)
    print(f"Training history saved to {filename}")

# 간단한 학습 루프
for model_name, base_model_class in [
    ("VGG16", VGG16),
    ("ResNet50", ResNet50),
    ("InceptionV3", InceptionV3),
    ("MobileNetV2", MobileNetV2)
]:
    print(f"Training {model_name}...")

    # 베이스 모델 로드
    base_model = base_model_class(weights='imagenet', include_top=False, input_shape=input_shape)
    base_model.trainable = False

    # 모델 생성
    model = create_model(base_model, num_classes=train_generator.num_classes)
    model.compile(optimizer=Adam(learning_rate=0.001), loss='categorical_crossentropy', metrics=['accuracy'])

    # 모델 학습
    history = model.fit(
        train_generator,
        validation_data=val_generator,
        epochs=3,  # 기본적으로 3 에포크만 실행
        verbose=1
    )

    # 히스토리 저장
    save_history(history, f"{model_name}_history.json")

    # 학습된 모델 저장
    model.save(f"{model_name}_basic_model.h5")
    print(f"{model_name} model saved!")

# Vision Transformer 학습
print("Training Vision Transformer (ViT)...")
vit_model = create_vit_model(num_classes=train_generator.num_classes, input_shape=input_shape)
vit_model.compile(optimizer=Adam(learning_rate=0.0001), loss="categorical_crossentropy", metrics=["accuracy"])

vit_history = vit_model.fit(
    train_generator,
    validation_data=val_generator,
    epochs=3,
    verbose=1
)

# 히스토리 저장
save_history(vit_history, "ViT_history.json")

vit_model.save("ViT_model.h5")
print("Vision Transformer model saved!")
